aws/resource_aws_msk_cluster.go

```diff
@@ -0,0 +1,307 @@
+			DataVolumeKMSKeyId: aws.String(encryptRestArn),
+		}
+	}
+
+	_, err := conn.CreateCluster(createOpts)
+	if err != nil {
+		return fmt.Errorf("Unable to create cluster: %s", err)
+	}
+
+	// No error, wait for ACTIVE state
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{kafka.ClusterStateCreating},
+		Target:     []string{kafka.ClusterStateActive},
+		Refresh:    clusterStateRefreshFunc(conn, cn),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	clusterRaw, err := stateConf.WaitForState()
+
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for MSK cluster (%s) to become active: %s",
+			cn, err)
+	}
+
+	c := clusterRaw.(*mskClusterState)
+	d.SetId(c.arn)
+	d.Set("arn", c.arn)
+	d.Set("status", c.status)
+	d.Set("zookeeper_connect", c.zookeeperConnect)
+	d.Set("bootstrap_brokers", c.bootstrapBrokers)
+
+	return nil
+}
+func resourceAwsMskClusterRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kafkaconn
+	name := d.Get("name").(string)
+
+	state, err := readMskClusterState(conn, name)
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			if awsErr.Code() == "ResourceNotFoundException" {
+				d.SetId("")
+				return nil
+			}
+			return fmt.Errorf("Error reading MSK cluster: \"%s\", code: \"%s\"", awsErr.Message(), awsErr.Code())
+		}
+		return err
+
+	}
+	d.SetId(state.arn)
+	d.Set("arn", state.arn)
+	d.Set("status", state.status)
+	d.Set("encrypt_rest_key", state.encryptRestKey)
+	d.Set("zookeeper_connect", state.zookeeperConnect)
+	d.Set("bootstrap_brokers", state.bootstrapBrokers)
+
+	return nil
+}
+
+func resourceAwsMskClusterDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).kafkaconn
+	arn := d.Get("arn").(string)
+	name := d.Get("name").(string)
+
+	_, err := conn.DeleteCluster(&kafka.DeleteClusterInput{
+		ClusterArn: aws.String(arn),
+	})
+	if err != nil {
+		return err
+	}
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{kafka.ClusterStateDeleting},
+		Target:     []string{"DESTROYED"},
+		Refresh:    clusterStateRefreshFunc(conn, name),
+		Timeout:    d.Timeout(schema.TimeoutDelete),
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for MSK Cluster (%s) to be destroyed: %s",
+			arn, err)
+	}
+
+	return nil
+}
+
+type mskClusterState struct {
+	arn               string
+	creationTimestamp int64
+	status            string
+	encryptRestKey    string
+	zookeeperConnect  string
+	bootstrapBrokers  string
+	brokerCount       int64
+	kafkaVersion      string
+}
+
+func readMskClusterState(conn *kafka.Kafka, name string) (*mskClusterState, error) {
+	listOpts := &kafka.ListClustersInput{
+		ClusterNameFilter: aws.String(name),
+	}
+
+	cluster_list, err := conn.ListClusters(listOpts)
```

Comment:
This should be replaced with [`DescribeCluster()`](https://docs.aws.amazon.com/sdk-for-go/api/service/kafka/#Kafka.DescribeCluster). We only need one direct result and discourage the below usage of creating our own errors when the AWS Go SDK already provides standard error handling.